### ISSUE TITLE
x/net/context: Documentation for Context.Err doesn't explicitly state what values to expect before Done() is closed

### ISSUE THREAD

**Issue by riannucci** [golang/go#14292]
In https://godoc.org/golang.org/x/net/context#Context, there are two possible readings for the documentation of Err:
1. Err() must return nil before Done() isn't closed, and atomically switch to returning some non-nil error when Done() IS closed
2. Err() is undefined before Done() is closed, and the only valid use for Err() is after blocking on Done()

Looking at the implementation, I suspect that 1 is the correct interpretation. I would propose that the documentation on this function be amended to read:

``` go
    // Err returns a non-nil error value after Done is closed.  Err returns
    // Canceled if the context was canceled or DeadlineExceeded if the
    // context's deadline passed.  If Done is not yet closed, Err must return nil.
    // No other values for Err are defined. After Done is closed, successive
    // calls to Err return the same value.
    Err() error
```

Note the added "If Done is not yet closed, Err must return nil." sentence.

**Comment by bcmills** [comment]
I would argue that (2) is correct.  There is no reason to call `Err` without `Done`: if nothing else, it encourages the writing of racy code.

(To poll whether a `Context` is done without blocking, you can use a `select` with a `default` case.)

**Comment by riannucci** [comment]
It's been a while, but if I recall, the reason to call `Err` might be something like:

```go
for ctx.Err() != nil {
  // do short operation that can't easily select on Done()
}
```

Otherwise this would need to look like

```go
outer: for {
  select {
    case <-ctx.Done():
      break outer
    default:
      // do short operation 
  }
}
```

I don't recall the exact specifics of why this came up, though. I don't see why expecting to be able to call Err() without Done() is inherently racy though? I suppose if you were checking Err() and then doing something with the context, that could be racy (sort-of... the context itself would have to be racy), but I don't see why that would be NOT the case in the event that you do something in the default clause of the select with the context object either:

```go
outer: for {
  select {
    case <-ctx.Done():
      break outer
    default:
      // context could actually be Done here anyway
  }
}
```

**Comment by bcmills** [comment]
A more concrete use-case would be helpful.  In my experience, more often than not the loops you're describing actually look like:

```go
for {
  select {
  case <-ctx.Done():
    return nil, ctx.Err()
  default:
  }
  // Do short operation that can't easily select on Done().
}
return someValue, nil
```

Calling `Err` without waiting on `Done` is racy if the particular implementation of the `Context` interface implements the `Err` method without adding synchronization beyond closing the `Done` channel.  There is currently no prohibition on such an implementation.  (The `cancelCtx` in the standard library uses a `sync.Mutex`, but that's more to guard the fast-path cancelers than `c.err` per se.)

The fundamental problem is this: if the implementation of `Err` doesn't add a hard sequence point to the program, then the compiler (or the CPU, for that matter!) can assume that its value never changes and optimize your loop to:
```go
for {
  // do short operation that can't easily select on Done()
}
```

But if `Err` does add a hard sequence point, it becomes much more expensive: for example, `cancelCtx.Err` acquires a `Mutex` lock, which means that if you have N CPUs all executing those loops simultaneously, your code will appear to execute many copies of the short operation in parallel but actually spend most of its time moving the mutex around from cache to cache.  (It's a similar problem to the one underlying https://github.com/golang/go/issues/17973.)

**Comment by riannucci** [comment]
I guess that's why I was hoping the documentation could be clarified 😄. It sounds like the documentation should say something along the lines of "It is undefined behavior to call Err() before blocking on Done()."

**Comment by riannucci** [comment]
(I was hoping that it would be reasonable to give Err() a meaning before Done() in order to make it safe, but it sounds like documenting this behavior is insufficient to make it reliable since other implementations of Context presumably don't couple these two.)

**Comment by riannucci** [comment]
(sorry for all the updates)

Would it make sense to have `govet` check for this misuse of Err?

**Comment by nya3jp** [comment]
I believe this issue can be closed now that #19856 was accepted and implemented.

**Comment by bcmills** [comment]
Duplicate of #19856
